### DISCUSSION TITLE
Creating all serverless functions taking 5+ mins

### DISCUSSION THREAD

**Discussion by akshitkrnagpal** [vercel/next.js#30137] [Help] [answered]
Hi everyone

I am facing this issue on Vercel that builds are taking 10+mins more than half of the time is for creating serverless functions. 

```
Created all serverless functions in: 6:13.097 (m:ss.mmm)
```

As I am not using getServerSideProps for any page I don't think serverless function needs to be created. That is confirmed from the `functions` tab of vercel because it shows "No Serverless Functions"

How can find the cause of this?

**Comment by ijjk** [comment, marked as answer]
Hi, can you test the latest version of Next.js `v11.1.3-canary.91` and add the below config to your `next.config.js`? We're working on improving the performance/reliability of this step in the latest version.  

```js
module.exports = {
  experimental: {
    outputFileTracing: true
  }
}
```

**Reply by akshitkrnagpal** [reply]
Hi @ijjk 

Yes. I tried the above changes and the build time went from 10mins to 4 mins. Creating all serverless functions reduced from 6 mins to 2 secs. But I noticed some of the pages were giving 500 error in deployed build (on vercel) whereas they working fine on dev and production build on my localhost. The issue seems to be if the page was to be generated by `fallback: "blocking"`, it was giving 500 error on Vercel. (Not during build time, but when visiting the page).

Do you know how this can be fixed?

**Reply by ijjk** [reply]
Can you share a link to the repo or to the deployment so we can investigate this further?

**Reply by akshitkrnagpal** [reply]
@ijjk I cannot share as it's a company project and we wouldn't like that to be shared in public right now. Can I DM you on twitter?

**Reply by ijjk** [reply]
Yes, feel free to DM https://twitter.com/_ijjk

**Reply by akshitkrnagpal** [reply]
@ijjk Thanks. DM'd you.

**Comment by rodilo** [comment]
hey @akshitkrnagpal @ijjk 
I think we have the same issue, I did some debugging and you can check it here
https://github.com/vercel/next.js/issues/30255
There's also a test repo I've used for debugging.

Basically, for us, it happens because of `i18n`. We don't have any pages with `getServerSideProps` or `getInitialProps`.
I tested some older versions of next, and this thing started happening after `10.1.*`

Our production app is a bit complex, and usually, it takes around 5-6 minutes to deploy on vercel. But after an upgrade to next 11, it takes as twice as long.

**Reply by akshitkrnagpal** [reply]
Hi @rodilo 

Have you tried the following suggestion from ijjk? What was your result?

> Hi, can you test the latest version of Next.js `v11.1.3-canary.91` and add the below config to your `next.config.js`? We're working on improving the performance/reliability of this step in the latest version.
> 
> ```js
> module.exports = {
>   experimental: {
>     outputFileTracing: true
>   }
> }
> ```

**Comment by rodilo** [comment]
@akshitkrnagpal yes, just now, I've got this error:

![image](https://user-images.githubusercontent.com/32063803/138716377-7d8314d6-8769-4860-a7de-f7a0c9e5cd1c.png)

**Reply by ijjk** [reply]
@rodilo which deployment are you seeing that error? The linked reproduction is showing a successful deploy without that error being present https://github.com/rodilo/next-18n-build-test/commit/d8e5858a39e48b2d4eaff3640c6068ead87a888f

**Reply by ijjk** [reply]
That error usually only occurs when you have incorrect build settings under your project's settings. It might say `Other` instead of Next.js or have an incorrect output directory.

**Reply by rodilo** [reply]
> @rodilo which deployment are you seeing that error? The linked reproduction is showing a successful deploy without that error being present https://github.com/rodilo/next-18n-build-test/commit/d8e5858a39e48b2d4eaff3640c6068ead87a888f

It happened on my private project (the one that takes 15 minutes to build). This test project run ok.

Yes, the build config on my private project is set to `Other`, I'll change that and try again. Thanks for the help 👍

**Comment by rodilo** [comment]
@ijjk 

OK, it works now, build runs in a reasonable time (~5 minutes instead of ~16-17).
I've found what was causing the issue with the build failing. 
I use this code for getting dynamic slugs:

```js
import matter from 'gray-matter'
import path from 'path'
import read from 'read-directory'

const raw = read.sync(path.join(process.cwd(), 'data/postItems'), { transform: matter })

```

and it works when I run `next dev` but it doesn't in `next build`. .
The problem is this `read-directory` package, so I've created my own function that does the same.

Other more important thing that it's not working are pages with `getServerSideProps`. I get 500 error and this when I try to check function logs on vercel:

![image](https://user-images.githubusercontent.com/32063803/138863959-e87b2a43-e643-4990-aa0f-fe5b9097d9ce.png)

You can check that on this preview link https://emmalife-p7vs2uf5e-creadi.vercel.app/feedback

Hope this helps. I'm looking forward to releasing this feature.

**Reply by ijjk** [reply]
The error you encountered viewing the logs should be corrected, it seems the 500 error is being caused by `@sentry/nextjs` specifically this file `/var/task/node_modules/@sentry/nextjs/dist/utils/instrumentServer.js:7:14` as it tries to import the entire `next` module which we don't allow since it would increase the deployment size drastically.

**Reply by rodilo** [reply]
> The error you encountered viewing the logs should be corrected, it seems the 500 error is being caused by `@sentry/nextjs` specifically this file `/var/task/node_modules/@sentry/nextjs/dist/utils/instrumentServer.js:7:14` as it tries to import the entire `next` module which we don't allow since it would increase the deployment size drastically.

And what about my test repo? I added `getServerSideProps` to one page, and I got the same error - 500, check it here:

https://next-18n-build-test-6g8yx03yk-shonely.vercel.app/second-page

In the functions log i get this:

![image](https://user-images.githubusercontent.com/32063803/138886021-7a55b736-2219-47c3-b12c-c4fba15c1c18.png)

**Reply by ijjk** [reply]
That issue looks related to https://github.com/isaachinman/next-i18next/issues/1202#issuecomment-947779838 you can work around that by adding a `fs.readdir(path.join(process.cwd(), 'public/locales')` to your `getServerSideProps` where `serverSideTranslations` is being used.

**Reply by rodilo** [reply]
Ah, those were two different things. 
@ijjk  thanks for your help, now I need to figure out what to do with `sentry` :)

**Reply by ijjk** [reply]
Could use `patch-package` and comment out [this call](https://github.com/getsentry/sentry-javascript/blob/9c25928327d7c0eaee1bdcf8a180b77e18c10653/packages/nextjs/src/index.server.ts#L85) also removing the import which should resolve it temporarily although would be good to open an issue for the package.
